File: dotNetTips.Utility.Standard.bak/IO/DirectoryHelper.cs
// ***********************************************************************
// Assembly         : dotNetTips.Utility.Standard
// Author           : David McCarter
// Created          : 01-16-2017
//
// Last Modified By : David McCarter
// Last Modified On : 01-20-2017
// ***********************************************************************
// <copyright file="DirectoryHelper.cs" company="dotNetTips.com">
//     Copyright (c) dotNetTips.com - McCarter Consulting. All rights reserved.
// </copyright>
// <summary></summary>
// ***********************************************************************
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace dotNetTips.Utility.Standard.IO
{
    /// <summary>
    /// Class DirectoryHelper.
    /// </summary>
    public static class DirectoryHelper
    {
        /// <summary>
        /// Applications the application data folder.
        /// </summary>
        /// <returns>System.String.</returns>
        public static string AppApplicationDataFolder()
        {
            //TODO: BLOG POST
            var userPath = Environment.GetEnvironmentVariable(
            RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "LOCALAPPDATA" : "Home");

            var assy = System.Reflection.Assembly.GetEntryAssembly();
            var companyName = assy.GetCustomAttributes<AssemblyCompanyAttribute>().FirstOrDefault();
            var path = System.IO.Path.Combine(userPath,companyName.Company);
            
            return path;
         }

        /// <summary>
        /// Loads the files.
        /// </summary>
        /// <param name="directories">The directories.</param>
        /// <param name="searchPattern">The search pattern.</param>
        /// <param name="searchOption">The search option.</param>
        /// <returns>IEnumerable(Of FileInfo).</returns>
        public static IEnumerable<FileInfo> LoadFiles(IEnumerable<DirectoryInfo> directories, string searchPattern, SearchOption searchOption)
        {

            var files = new List<FileInfo>();

            //TODO: BLOG POST
            Parallel.ForEach(directories, (directory) =>
            {
                if ((directory.Exists))
                {
                    var foundFiles = directory.EnumerateFiles(searchPattern, searchOption);
                    lock (files)
                    {
                        files.AddRange(foundFiles);
                    }
                }
            });

            return files.Distinct().AsEnumerable();

        }

        /// <summary>
        /// delete directory as an asynchronous operation.
        /// </summary>
        /// <param name="directory">The directory.</param>
        /// <returns>Task&lt;System.Boolean&gt;.</returns>
        /// <exception cref="System.ArgumentNullException">directory</exception>
        public static async Task<bool> DeleteDirectoryAsync(DirectoryInfo directory)
        {
            if (directory == null)
            {
                throw new ArgumentNullException(nameof(directory));
            }

            if (directory.Exists)
            {
                await Task.Factory.StartNew(() =>
                {
                    directory.Delete(recursive: true);
                    return true;
                }).ConfigureAwait(true);
            }
            return false;
        }
    }
}

File: dotNetTips.Utility/IO/DirectoryHelper.vb
' ***********************************************************************
' Assembly         : dotNetTips.Utility
' Author           : David McCarter
' Created          : 04-15-2016
'
' Last Modified By : David McCarter
' Last Modified On : 06-24-2017
' ***********************************************************************
' <copyright file="DirectoryHelper.vb" company="NicheWare - David McCarter">
'     NicheWare - David McCarter
' </copyright>
' <summary></summary>
' *************************************************************************
Imports System.Collections.Generic
Imports System.IO
Imports System.Threading.Tasks

Namespace IO
    ''' <summary>
    ''' Class DirectoryHelper.
    ''' </summary>
    Public Module DirectoryHelper
        ''' <summary>
        ''' Applications the application data folder.
        ''' </summary>
        ''' <returns>System.String.</returns>
        Public Function AppApplicationDataFolder() As String
            Dim path = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), My.Application.Info.CompanyName.Trim)

            Return path

        End Function

        ''' <summary>
        ''' Loads the files.
        ''' </summary>
        ''' <param name="directories">The directories.</param>
        ''' <param name="searchPattern">The search pattern.</param>
        ''' <param name="searchOption">The search option.</param>
        ''' <param name="throwException">if set to <c>true</c> [throw exception].</param>
        ''' <returns>IEnumerable(Of FileInfo).</returns>
        Public Function LoadFiles(ByVal directories As IEnumerable(Of DirectoryInfo), searchPattern As String, searchOption As SearchOption, throwException As Boolean) As IEnumerable(Of FileInfo)

            Dim files = New List(Of FileInfo)

            For Each directory In directories.AsParallel
                Try
                    SearchForFiles(searchPattern, searchOption, files, directory)
                Catch securityEx As UnauthorizedAccessException
                    If throwException Then
                        Throw
                    End If
                End Try
            Next

            Return files.Distinct.AsEnumerable

        End Function

        ''' <summary>
        ''' Searches for files.
        ''' </summary>
        ''' <param name="searchPattern">The search pattern.</param>
        ''' <param name="searchOption">The search option.</param>
        ''' <param name="files">The files.</param>
        ''' <param name="directory">The directory.</param>
        Private Sub SearchForFiles(searchPattern As String, searchOption As SearchOption, files As List(Of FileInfo), directory As DirectoryInfo)
            If (directory.Exists) Then
                Dim foundFiles = directory.EnumerateFiles(searchPattern, searchOption)
                SyncLock files
                    files.AddRange(foundFiles)
                End SyncLock
            End If
        End Sub

        ''' <summary>
        ''' Loads the files.
        ''' </summary>
        ''' <param name="directories">The directories.</param>
        ''' <param name="searchPattern">The search pattern.</param>
        ''' <param name="searchOption">The search option.</param>
        ''' <returns>IEnumerable(Of FileInfo).</returns>
        Public Function LoadFiles(ByVal directories As IEnumerable(Of DirectoryInfo), searchPattern As String, searchOption As SearchOption) As IEnumerable(Of FileInfo)

            Dim files = New List(Of FileInfo)

            For Each directory As DirectoryInfo In directories.AsParallel
                If (directory.Exists) Then
                    Dim foundFiles = directory.EnumerateFiles(searchPattern, searchOption)
                    SyncLock files
                        files.AddRange(foundFiles)
                    End SyncLock
                End If
            Next

            Return files.Distinct.AsEnumerable

        End Function

        ''' <summary>
        ''' delete directory as an asynchronous operation.
        ''' </summary>
        ''' <param name="directory">The directory.</param>
        ''' <returns>Task(Of System.Boolean).</returns>
        ''' <exception cref="System.ArgumentNullException">directory</exception>
        Public Async Function DeleteDirectoryAsync(directory As DirectoryInfo) As Task(Of Boolean)
            If directory Is Nothing Then
                Throw New ArgumentNullException(NameOf(directory))
            End If

            If directory.Exists Then
                Await Task.Factory.StartNew(Function()
                                                directory.Delete(recursive:=True)
                                                Return True
                                            End Function).ConfigureAwait(True)
            End If
            Return False
        End Function
    End Module
End Namespace